mempool/cat/reactor.go

```diff
@@ -0,0 +1,448 @@
+	requests *requestScheduler
+}
+
+type ReactorOptions struct {
+	// ListenOnly means that the node will never broadcast any of the transactions that
+	// it receives. This is useful for keeping transactions private
+	ListenOnly bool
+
+	// MaxTxSize is the maximum size of a transaction that can be received
+	MaxTxSize int
+
+	// MaxGossipDelay is the maximum allotted time that the reactor expects a transaction to
+	// arrive before issuing a new request to a different peer
+	MaxGossipDelay time.Duration
+}
+
+func (opts *ReactorOptions) VerifyAndComplete() error {
+	if opts.MaxTxSize == 0 {
+		opts.MaxTxSize = cfg.DefaultMempoolConfig().MaxTxBytes
+	}
+
+	if opts.MaxGossipDelay == 0 {
+		opts.MaxGossipDelay = defaultGossipDelay
+	}
+
+	if opts.MaxTxSize < 0 {
+		return fmt.Errorf("max tx size (%d) cannot be negative", opts.MaxTxSize)
+	}
+
+	if opts.MaxGossipDelay < 0 {
+		return fmt.Errorf("max gossip delay (%d) cannot be negative", opts.MaxGossipDelay)
+	}
+
+	return nil
+}
+
+// NewReactor returns a new Reactor with the given config and mempool.
+func NewReactor(mempool *TxPool, opts *ReactorOptions) (*Reactor, error) {
+	err := opts.VerifyAndComplete()
+	if err != nil {
+		return nil, err
+	}
+	memR := &Reactor{
+		opts:     opts,
+		mempool:  mempool,
+		ids:      newMempoolIDs(),
+		requests: newRequestScheduler(opts.MaxGossipDelay, defaultGlobalRequestTimeout),
+	}
+	memR.BaseReactor = *p2p.NewBaseReactor("Mempool", memR)
+	return memR, nil
+}
+
+// SetLogger sets the Logger on the reactor and the underlying mempool.
+func (memR *Reactor) SetLogger(l log.Logger) {
+	memR.Logger = l
+}
+
+// OnStart implements p2p.BaseReactor.
+func (memR *Reactor) OnStart() error {
+	if memR.opts.ListenOnly {
+		memR.Logger.Info("Tx broadcasting is disabled")
+		return nil
+	}
+	go func() {
+		for {
+			select {
+			case <-memR.Quit():
+				return
+
+			// listen in for any newly verified tx via RFC, then immediately
+			// broadcasts it to all connected peers.
+			case nextTxKey := <-memR.mempool.next():
+				wtx := memR.mempool.store.get(nextTxKey)
+				// tx may have been removed after it was added to the broadcast queue
+				// thus we just skip over it
+				if wtx != nil {
+					memR.broadcastNewTx(wtx)
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+func (memR *Reactor) OnStop() {
+	// stop all the timers tracking outbound requests
+	memR.requests.Close()
+}
+
+// GetChannels implements Reactor by returning the list of channels for this
+// reactor.
+func (memR *Reactor) GetChannels() []*p2p.ChannelDescriptor {
+	largestTx := make([]byte, memR.opts.MaxTxSize)
+	txMsg := protomem.Message{
+		Sum: &protomem.Message_Txs{
+			Txs: &protomem.Txs{Txs: [][]byte{largestTx}},
+		},
+	}
+
+	stateMsg := protomem.Message{
+		Sum: &protomem.Message_SeenTx{
+			SeenTx: &protomem.SeenTx{
+				TxKey: make([]byte, tmhash.Size),
+			},
+		},
+	}
+
+	return []*p2p.ChannelDescriptor{
+		{
+			ID:                  mempool.MempoolChannel,
+			Priority:            6,
```

Comment:
is there a list of priorities  for all of the reactors and the reasoning behind them anywhere? The other mempools have the mempool channel listed at priority 5, I'm guessing that this was increased simply to favor sending txs rather than mempool state?

[for other reviewers to compare]
```go
### blockchain (block sync)

{
	ID:                  BlockchainChannel,
	Priority:            5,
},

### evidence

{
	ID:                  EvidenceChannel,
	Priority:            6,
}

### mempool v1 and v0

{
	ID:                  mempool.MempoolChannel,
	Priority:            5,
},

### statesync

{
	ID:                  SnapshotChannel,
	Priority:            5,,
},
{
	ID:                  ChunkChannel,
	Priority:            3,
},

### Consensus
{
	ID:                  StateChannel,
	Priority:            6,
},
{
	ID:                  DataChannel,
	Priority:            10,
},
{
	ID:                  VoteChannel,
	Priority:            7,
},
{
	ID:                  VoteSetBitsChannel,
	Priority:            1,
},
```

[general tendermint p2p question]
do we know if priorities actually make that big of a difference?

Reply:
There hasn't been any empirical study on this that I know of. I think the affect was more noticeable in v0.35 because priority was implemented both for outbound messages as well as inbound messages. In v0.34 priority only affects outgoing messages. The logic there is quite nuanced and difficult to reason about but I think channels which have a high priority and haven't sent something recently get priority for their messages. There's some flowrate library which manages it. I remember there was some conversations in the past about whether it is necessary - especially in context of the migration to libp2p